FFT execution failure in ROCm 4.2.0

I'm investigating a number of testing failures of `cupy.fft` under ROCm 4.2.0. Some fft execution fails with return code `CUFFT_EXEC_FAILED`, that is actually `HIPFFT_EXEC_FAILED` in hipFFT, and it seems that rocFFT internally called returns `rocfft_status_invalid_work_buffer` there.

```
________________________ TestPlanCache.test_LRU_cache4 _________________________

self = <cupy_tests.fft_tests.test_cache.TestPlanCache testMethod=test_LRU_cache4>

    def test_LRU_cache4(self):
        # test if fetching the plan will reorder it to the top
        cache = config.get_plan_cache()
        assert cache.get_curr_size() == 0 <= cache.get_size()

        # this creates a Plan1d
        a = testing.shaped_random((10,), cupy, cupy.float32)
        cupy.fft.fft(a)
        assert cache.get_curr_size() == 1 <= cache.get_size()

        # this creates a PlanNd
        a = testing.shaped_random((10, 20), cupy, cupy.float32)
>       cupy.fft.fftn(a)

tests/cupy_tests/fft_tests/test_cache.py:120:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
cupy/fft/_fft.py:772: in fftn
    return func(a, s, axes, norm, cufft.CUFFT_FORWARD)
cupy/fft/_fft.py:614: in _fftn
    a = _exec_fftn(a, direction, value_type, norm=norm, axes=axes_sorted,
cupy/fft/_fft.py:554: in _exec_fftn
    plan.fft(a, out, direction)
cupy/cuda/cufft.pyx:838: in cupy.cuda.cufft.PlanNd.fft
    execC2C(plan, a.data.ptr, out.data.ptr, direction)
cupy/cuda/cufft.pyx:1100: in cupy.cuda.cufft.execC2C
    check_result(result)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   raise CuFFTError(result)
E   cupy.cuda.cufft.CuFFTError: CUFFT_EXEC_FAILED

cupy/cuda/cufft.pyx:169: CuFFTError
=========================== short test summary info ============================
FAILED tests/cupy_tests/fft_tests/test_cache.py::TestPlanCache::test_LRU_cache4
```

--- a/.pfnci/config.pbtxt
+++ b/.pfnci/config.pbtxt
@@ -145,6 +145,7 @@ configs {
       include_dot_git: true
     }
     environment_variables { key: "ROCM_REPO" value: "4.0.1" }
+    environment_variables { key: "ROCM_PACKAGES" value: "rocm-dev hipblas hipsparse rocsparse rocrand rocthrust rocsolver rocfft hipcub rocprim rccl" }
     command: "bash .pfnci/script.sh py37.rocm-4-0"
   }
 }
@@ -163,6 +164,7 @@ configs {
       include_dot_git: true
     }
     environment_variables { key: "ROCM_REPO" value: "debian" }
+    environment_variables { key: "ROCM_PACKAGES" value: "rocm-dev hipblas hipfft hipsparse rocsparse rocrand rocthrust rocsolver rocfft hipcub rocprim rccl" }
     command: "bash .pfnci/script.sh py37.rocm-latest"
   }
 }

--- a/.pfnci/script.sh
+++ b/.pfnci/script.sh
@@ -79,7 +79,7 @@ main() {
       apt-get autoremove -qqy
 
       apt update -qqy
-      apt install rocm-dev hipblas hipsparse rocsparse rocrand rocthrust rocsolver rocfft hipcub rocprim rccl -qqy
+      apt install rocm-dev hipblas hipfft hipsparse rocsparse rocrand rocthrust rocsolver rocfft hipcub rocprim rccl -qqy
       export HCC_AMDGPU_TARGET=gfx900
       export ROCM_HOME=/opt/rocm
       export CUPY_INSTALL_USE_HIP=1

--- a/.pfnci/script.sh
+++ b/.pfnci/script.sh
@@ -79,7 +79,11 @@ main() {
       apt-get autoremove -qqy
 
       apt update -qqy
-      apt install rocm-dev hipblas hipfft hipsparse rocsparse rocrand rocthrust rocsolver rocfft hipcub rocprim rccl -qqy
+      apt install rocm-dev hipblas hipsparse rocsparse rocrand rocthrust rocsolver rocfft hipcub rocprim rccl -qqy
+      # Also install hipfft for ROCm 4.1 or later
+      if [ "${TARGET}" == cupy.py37.rocm-latest ]; then
+          apt install hipfft -qqy
+      fi
       export HCC_AMDGPU_TARGET=gfx900
       export ROCM_HOME=/opt/rocm
       export CUPY_INSTALL_USE_HIP=1

--- a/.pfnci/script.sh
+++ b/.pfnci/script.sh
@@ -79,11 +79,7 @@ main() {
       apt-get autoremove -qqy
 
       apt update -qqy
-      apt install rocm-dev hipblas hipsparse rocsparse rocrand rocthrust rocsolver rocfft hipcub rocprim rccl -qqy
-      # Also install hipfft for ROCm 4.1 or later
-      if [ "${TARGET}" == py37.rocm-latest ]; then
-          apt install hipfft -qqy
-      fi
+      apt install $ROCM_PACKAGES -qqy
       export HCC_AMDGPU_TARGET=gfx900
       export ROCM_HOME=/opt/rocm
       export CUPY_INSTALL_USE_HIP=1

--- a/.pfnci/script.sh
+++ b/.pfnci/script.sh
@@ -81,7 +81,7 @@ main() {
       apt update -qqy
       apt install rocm-dev hipblas hipsparse rocsparse rocrand rocthrust rocsolver rocfft hipcub rocprim rccl -qqy
       # Also install hipfft for ROCm 4.1 or later
-      if [ "${TARGET}" == cupy.py37.rocm-latest ]; then
+      if [ "${TARGET}" == py37.rocm-latest ]; then
           apt install hipfft -qqy
       fi
       export HCC_AMDGPU_TARGET=gfx900

--- a/cupy_setup_build.py
+++ b/cupy_setup_build.py
@@ -546,6 +546,8 @@ def preconfigure_modules(compiler, settings):
                     rocm_path = build.get_rocm_path()
                     inc_path = os.path.join(rocm_path, 'hipfft', 'include')
                     settings['include_dirs'].insert(0, inc_path)
+                    lib_path = os.path.join(rocm_path, 'hipfft', 'lib')
+                    settings['library_dirs'].insert(0, lib_path)
                 canonicalize_hip_libraries(hip_version, module['libraries'])
 
         print('')

--- a/cupy_setup_build.py
+++ b/cupy_setup_build.py
@@ -115,6 +115,7 @@ if use_hip:
         'libraries': [
             'amdhip64',  # was hiprtc and hip_hcc before ROCm 3.8.0
             'hipblas',
+            ('hipfft', lambda hip_version: hip_version >= 401),
             'hiprand',
             'rocfft',
             'roctx64',
@@ -452,6 +453,20 @@ def check_library(compiler, includes=(), libraries=(),
     return True
 
 
+def canonicalize_hip_libraries(hip_version, libraries):
+    def ensure_tuple(x):
+        return x if isinstance(x, tuple) else (x, None)
+    new_libraries = []
+    for library in libraries:
+        lib_name, pred = ensure_tuple(library)
+        if pred is None:
+            new_libraries.append(lib_name)
+        elif pred(hip_version):
+            new_libraries.append(lib_name)
+    libraries.clear()
+    libraries.extend(new_libraries)
+
+
 def preconfigure_modules(compiler, settings):
     """Returns a list of modules buildable in given environment and settings.
 
@@ -519,6 +534,19 @@ def preconfigure_modules(compiler, settings):
             lib_path = os.path.join(cugraph_path, 'lib')
             if os.path.exists(lib_path):
                 settings['library_dirs'].append(lib_path)
+
+        # In ROCm 4.1 and later, we need to use the independent version of
+        # hipfft as well as rocfft. We configure the lists of include
+        # directories and libraries to link here depending on ROCm version
+        # before the configuration process following.
+        if use_hip and module['name'] == 'cuda':
+            if module['check_method'](compiler, settings):
+                hip_version = module['version_method']()
+                if hip_version >= 401:
+                    rocm_path = build.get_rocm_path()
+                    inc_path = os.path.join(rocm_path, 'hipfft', 'include')
+                    settings['include_dirs'].insert(0, inc_path)
+                canonicalize_hip_libraries(hip_version, module['libraries'])
 
         print('')
         print('-------- Configuring Module: {} --------'.format(

